tracee-rules/engine/engine.go

```diff
@@ -69,7 +69,6 @@ func NewEngine(sigs []types.Signature, sources EventSources, output chan types.F
 func (engine *Engine) Start(done chan bool) {
 	go engine.consumeSources(done)
 	for s, c := range engine.signatures {
-		defer close(c)
```

Comment:
Here, defer close(c) inside the loop don’t get executed until the Start func ends - not when each step of for loop ends.

Reply:
what you say is correct, but it's not a bug though, is it?

Reply:
Well, yes and no 🤓 Start is run in a different thread than consumeSources. I temporarily printed goroutine id and reproduced this error with by intercepting it with ctrl+C

```console
$ cat tracee_output.gob| go run . --rules-dir dist/rules --input-tracee=file:stdin --input-tracee=format:gob
Loaded signature(s):  [TRC-1 TRC-2 TRC-3 TRC-4 TRC-5 TRC-6 TRC-7]
consumeSources goroutine 82
^CStart() is about to return
closing sig channel 0xc00010e540 in goroutine 1
closing sig channel 0xc00010e540 in goroutine 1
closing sig channel 0xc00010e540 in goroutine 1
closing sig channel 0xc00010e540 in goroutine 1
closing sig channel 0xc00010e540 in goroutine 1
closing sig channel 0xc00010e540 in goroutine 1
closing sig channel 0xc00010e540 in goroutine 1
panic: send on closed channel

goroutine 82 [running]:
github.com/aquasecurity/tracee/tracee-rules/engine.(*Engine).consumeSources(0xc0004ae620, 0xc0001a6000)
	/Users/dpacak/go/src/github.com/aquasecurity/tracee/tracee-rules/engine/engine.go:138 +0x33c
created by github.com/aquasecurity/tracee/tracee-rules/engine.(*Engine).Start
	/Users/dpacak/go/src/github.com/aquasecurity/tracee/tracee-rules/engine/engine.go:89 +0x65
exit status 2
```

From this listing you can see that we're closing channels in goroutine `1` but goroutine `82` is still trying to send an event on the closed channel.

Reply:
It's even more interesting to run this program with the -race flag and reproduce the panic

```console
$ cat tracee_output.gob| go run -race . --rules-dir dist/rules --input-tracee=file:stdin --input-tracee=format:gob
2021/07/02 15:17:54 error opening plugin builtin.so: plugin.Open("dist/rules/builtin"): plugin was built with a different version of package runtime
Loaded signature(s):  [TRC-2 TRC-3 TRC-4 TRC-5 TRC-6 TRC-7]
consumeSources goroutine 53

*** Detection ***
Time: 2021-07-02T13:18:03Z
Signature ID: TRC-2
Signature: Anti-Debugging
Data: map[]
Command: strace
Hostname: tracee
^CStart() is about to return
closing sig channel 0xc000110660 in goroutine 1
==================
WARNING: DATA RACE
Write at 0x00c000110670 by main goroutine:
  runtime.closechan()
      /usr/local/Cellar/go/1.16.5/libexec/src/runtime/chan.go:355 +0x0
  github.com/aquasecurity/tracee/tracee-rules/engine.(*Engine).Start.func1()
      /Users/dpacak/go/src/github.com/aquasecurity/tracee/tracee-rules/engine/engine.go:93 +0xe4
  github.com/aquasecurity/tracee/tracee-rules/engine.(*Engine).Start()
      /Users/dpacak/go/src/github.com/aquasecurity/tracee/tracee-rules/engine/engine.go:107 +0x23f
  main.main.func1()
      /Users/dpacak/go/src/github.com/aquasecurity/tracee/tracee-rules/main.go:87 +0x964
  github.com/urfave/cli/v2.(*App).RunContext()
      /Users/dpacak/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/app.go:322 +0xabe
  github.com/urfave/cli/v2.(*App).Run()
      /Users/dpacak/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/app.go:224 +0xb1c
  main.main()
      /Users/dpacak/go/src/github.com/aquasecurity/tracee/tracee-rules/main.go:125 +0xa95

Previous read at 0x00c000110670 by goroutine 45:
  runtime.chansend()
      /usr/local/Cellar/go/1.16.5/libexec/src/runtime/chan.go:158 +0x0
  github.com/aquasecurity/tracee/tracee-rules/engine.(*Engine).consumeSources()
      /Users/dpacak/go/src/github.com/aquasecurity/tracee/tracee-rules/engine/engine.go:138 +0x45a

Goroutine 45 (running) created at:
  github.com/aquasecurity/tracee/tracee-rules/engine.(*Engine).Start()
      /Users/dpacak/go/src/github.com/aquasecurity/tracee/tracee-rules/engine/engine.go:89 +0x76
  main.main.func1()
      /Users/dpacak/go/src/github.com/aquasecurity/tracee/tracee-rules/main.go:87 +0x964
  github.com/urfave/cli/v2.(*App).RunContext()
      /Users/dpacak/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/app.go:322 +0xabe
  github.com/urfave/cli/v2.(*App).Run()
      /Users/dpacak/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/app.go:224 +0xb1c
  main.main()
      /Users/dpacak/go/src/github.com/aquasecurity/tracee/tracee-rules/main.go:125 +0xa95
==================
closing sig channel 0xc000110600 in goroutine 1
closing sig channel 0xc0001105a0 in goroutine 1
closing sig channel 0xc000110540 in goroutine 1
closing sig channel 0xc0001104e0 in goroutine 1
closing sig channel 0xc000110480 in goroutine 1
Found 1 data race(s)
exit status 66
```